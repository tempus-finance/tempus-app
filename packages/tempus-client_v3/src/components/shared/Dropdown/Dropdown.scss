$dropdown-height: 40px;
$dropdown-width: 160px;
$dropdown-border-size: 1px;
$dropdown-border: $dropdown-border-size solid var(--dropdownBorder);
$dropdown-border-radius: 4px;
$dropdown-popup-offset: 7px;

.tc__dropdown {
  position: relative;
  width: $dropdown-width;
  height: $dropdown-height;

  .tc__tooltip-wrapper-anchor {
    width: 100%;
    height: 100%;
    padding: 8px;
    display: flex;
    align-items: center;
    justify-content: space-between;
    background-color: var(--dropdownBackground);
    border: $dropdown-border;
    box-sizing: border-box;
    border-radius: $dropdown-border-radius;
    box-shadow: var(--downWeakShadow);

    &:hover {
      background-color: var(--dropdownBackgroundHighlight);
    }

    &[data-selected='true'] {
      background-color: var(--dropdownBackgroundSelected);
    }
  }

  .tc__tooltip-wrapper {
    width: 100%;
    height: 100%;

    .tc__tooltip {
      width: var(--anchorWidth);
      box-sizing: border-box;

      &::before {
        content: initial;
      }

      &::after {
        content: initial;
      }
    }
  }
}

.tc__dropdown__popup {
  display: flex;
  flex-direction: column;
  align-items: flex-start;
  padding: 12px;
  box-sizing: border-box;
  cursor: default;
}

.tc__dropdown-popup-title {
  margin-bottom: 8px;
}
