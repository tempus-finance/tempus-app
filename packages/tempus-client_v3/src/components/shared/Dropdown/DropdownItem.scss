.tc__dropdownItem {
  display: flex;
  align-items: center;
  justify-content: space-between;
  height: 24px;
  width: 100%;

  &:not(:last-child) {
    margin-bottom: 16px;
  }
}

.tc__dropdownItem__selectable {
  align-items: flex-start;
  position: relative;

  .tc__btn {
    width: 100%;
    height: 100%;
    position: relative;
    text-align: start;
  }

  .tc__dropdownItem__icon-container {
    height: 100%;
    position: absolute;
    top: 0;
    right: 0;
    display: flex;
    align-items: center;
  }
}
