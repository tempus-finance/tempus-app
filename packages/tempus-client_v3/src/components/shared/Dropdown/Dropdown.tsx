import { FC, memo, MouseEvent, useCallback, useState } from 'react';
import ButtonWrapper from '../ButtonWrapper';
import Icon from '../Icon';
import Typography from '../Typography';

import './Dropdown.scss';

export interface DropdownProps {
  label: string;
  popupTitle?: string;
}

const Dropdown: FC<DropdownProps> = props => {
  const { label, popupTitle = '', children } = props;

  const [open, setOpen] = useState<boolean>(false);

  const onClick = useCallback(() => {
    setOpen(prevState => !prevState);
  }, []);

  const onPopupClick = useCallback((event: MouseEvent<HTMLDivElement>) => {
    // We don't want to close popup when user clicks on the popup itself
    event.stopPropagation();
  }, []);

  return (
    <ButtonWrapper className="tc__dropdown" onClick={onClick} selected={open}>
      <Typography variant="body-secondary">{label}</Typography>
      <Icon type={open ? 'up-chevron' : 'down-chevron'} size={12} />

      {/* Popup container */}
      {open && (
        <div className="tc__dropdown__popup" onClick={onPopupClick}>
          {popupTitle && (
            <div className="tc__dropdown-popup-title">
              <Typography variant="body-tertiary" weight="bold">
                {popupTitle}
              </Typography>
            </div>
          )}
          {/* Dropdown list items */}
          {children}
        </div>
      )}
    </ButtonWrapper>
  );
};
export default memo(Dropdown);
