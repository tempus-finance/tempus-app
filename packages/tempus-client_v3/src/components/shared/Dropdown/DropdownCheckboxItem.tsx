import { ChangeEvent, FC, useCallback, useState } from 'react';
import Checkbox from '../Checkbox';
import Icon, { IconType } from '../Icon';

import './DropdownItem.scss';

export interface DropdownCheckboxItemProps {
  label: string;
  onChange: (checked: boolean, label: string) => void;
  icon?: IconType;
}

const DropdownCheckboxItem: FC<DropdownCheckboxItemProps> = props => {
  const { label, icon, onChange } = props;

  const [checked, setChecked] = useState<boolean>(false);

  const onCheckboxToggle = useCallback(
    (event: ChangeEvent<HTMLInputElement>) => {
      setChecked(event.target.checked);

      onChange(event.target.checked, label);
    },
    [label, onChange],
  );

  return (
    <div className="tc__dropdownItem">
      <Checkbox checked={checked} label={label} onChange={onCheckboxToggle} />
      {icon && <Icon type={icon} size={12} />}
    </div>
  );
};
export default DropdownCheckboxItem;
