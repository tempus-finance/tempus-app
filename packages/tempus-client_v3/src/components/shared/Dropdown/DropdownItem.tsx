import { ChangeEvent, FC, useCallback, useState } from 'react';
import Checkbox from '../Checkbox';
import Icon, { IconVariant } from '../Icon';
import Typography from '../Typography';

import './DropdownItem.scss';

export interface DropdownItemProps {
  label: string;
  onChange: (checked: boolean, label: string) => void;
  checkbox?: boolean;
  icon?: IconVariant;
}

const DropdownItem: FC<DropdownItemProps> = props => {
  const { label, checkbox = false, icon, onChange } = props;

  const [checked, setChecked] = useState<boolean>(false);

  const onCheckboxToggle = useCallback(
    (event: ChangeEvent<HTMLInputElement>) => {
      setChecked(event.target.checked);

      onChange(event.target.checked, label);
    },
    [label, onChange],
  );

  return (
    <div className="tc__dropdownItem">
      {checkbox && <Checkbox checked={checked} label={label} onChange={onCheckboxToggle} />}
      {!checkbox && (
        <Typography variant="body-secondary" weight={checked ? 'bold' : 'regular'}>
          {label}
        </Typography>
      )}
      {icon && <Icon variant={icon} size={12} />}
    </div>
  );
};
export default DropdownItem;
