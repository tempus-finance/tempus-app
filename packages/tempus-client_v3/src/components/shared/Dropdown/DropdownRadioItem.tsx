import { ReactElement, useCallback } from 'react';
import { Radio } from '../Radio';
import Icon, { IconVariant } from '../Icon';

import './DropdownItem.scss';

export interface DropdownRadioItemProps<T> {
  label: string;
  value: T;
  checked?: boolean;
  onChange: (label: T) => void;
  icon?: IconVariant;
}

function DropdownRadioItem<T>(
  props: DropdownRadioItemProps<T | string>,
): ReactElement<DropdownRadioItemProps<T | string>> {
  const { label, value, checked = false, icon, onChange } = props;

  const onRadioToggle = useCallback(() => {
    onChange(value);
  }, [onChange, value]);

  return (
    <div className="tc__dropdownItem">
      <Radio checked={checked} label={label} onChange={onRadioToggle} />
      {icon && <Icon variant={icon} size="tiny" />}
    </div>
  );
}
export default DropdownRadioItem;
