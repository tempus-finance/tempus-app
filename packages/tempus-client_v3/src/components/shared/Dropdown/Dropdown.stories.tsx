import { ComponentStory, ComponentMeta } from '@storybook/react';
import { useState } from 'react';
import Dropdown from './Dropdown';
import DropdownCheckboxItem from './DropdownCheckboxItem';
import DropdownSelectableItem from './DropdownSelectableItem';
import DropdownSelectableItemIcon from './DropdownSelectableItemIcon';
import DropdownSelector from './DropdownSelector';

export default {
  title: 'Dropdown',
  component: Dropdown,
  argTypes: {
    label: {
      control: {
        type: 'text',
      },
    },
    popupTitle: {
      control: {
        type: 'text',
      },
    },
  },
} as ComponentMeta<typeof Dropdown>;

const style = {
  display: 'flex',
  alignItems: 'center',
  justifyContent: 'center',
  height: '100px',
};

const SingleChoiceTemplate: ComponentStory<typeof Dropdown> = args => {
  const [selectedValue, setSelectedValue] = useState('a');

  return (
    <div style={style}>
      <DropdownSelector {...args} selectedValue={selectedValue} onSelect={setSelectedValue}>
        <DropdownSelectableItem label="Option A" value="a" />
        <DropdownSelectableItem label="Option B" value="b" />
        <DropdownSelectableItem label="Option C" value="c" />
      </DropdownSelector>
    </div>
  );
};

export const SingleChoice = SingleChoiceTemplate.bind({});
SingleChoice.args = {
  label: 'Filter',
  popupTitle: 'Title',
};

const SingleChoiceWithIconsTemplate: ComponentStory<typeof Dropdown> = args => {
  const [selectedValue, setSelectedValue] = useState('a1');

  return (
    <div style={style}>
      <DropdownSelector {...args} selectedValue={selectedValue} onSelect={setSelectedValue}>
        <DropdownSelectableItem label="Option A">
          <DropdownSelectableItemIcon icon="up-arrow-thin" value="a1" />
          <DropdownSelectableItemIcon icon="down-arrow-thin" value="a2" />
        </DropdownSelectableItem>
        <DropdownSelectableItem label="Option B" value="b" />
        <DropdownSelectableItem label="Option C" value="c" />
      </DropdownSelector>
    </div>
  );
};

export const SingleChoiceWithIcons = SingleChoiceWithIconsTemplate.bind({});
SingleChoiceWithIcons.args = {
  label: 'Filter',
  popupTitle: 'Title',
};

const MultipleChoiceTemplate: ComponentStory<typeof Dropdown> = args => (
  <div style={style}>
    <Dropdown {...args}>
      <DropdownCheckboxItem label="Active" onChange={() => {}} />
      <DropdownCheckboxItem label="Matured" onChange={() => {}} />
      <DropdownCheckboxItem label="Inactive" onChange={() => {}} />
    </Dropdown>
  </div>
);

export const MultipleChoice = MultipleChoiceTemplate.bind({});
MultipleChoice.args = {
  label: 'Filter',
  popupTitle: 'Title',
};

const RightSideIconTemplate: ComponentStory<typeof Dropdown> = args => (
  <div style={style}>
    <Dropdown {...args}>
      <DropdownCheckboxItem label="Active" onChange={() => {}} icon="up-arrow-thin" />
      <DropdownCheckboxItem label="Matured" onChange={() => {}} />
      <DropdownCheckboxItem label="Inactive" onChange={() => {}} />
    </Dropdown>
  </div>
);

export const RightSideIcon = RightSideIconTemplate.bind({});
RightSideIcon.args = {
  label: 'Filter',
  popupTitle: 'Title',
};
