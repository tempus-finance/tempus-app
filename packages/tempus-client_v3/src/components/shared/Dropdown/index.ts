import Dropdown from './Dropdown';
import DropdownCheckboxItem from './DropdownCheckboxItem';
import DropdownSelectableItem from './DropdownSelectableItem';
import DropdownSelector from './DropdownSelector';

export { Dropdown, DropdownCheckboxItem, DropdownSelectableItem, DropdownSelector };
