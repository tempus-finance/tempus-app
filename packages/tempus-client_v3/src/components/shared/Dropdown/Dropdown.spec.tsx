import { fireEvent, render } from '@testing-library/react';
import Dropdown, { DropdownProps } from './Dropdown';

const defaultProps: DropdownProps = {
  label: 'Filter',
  popupTitle: 'Title',
};

const subject = (props: DropdownProps) => render(<Dropdown {...props} />);

describe('Dropdown', () => {
  it('renders a button element', () => {
    const { getByRole } = subject(defaultProps);

    const result = getByRole('button');

    expect(result).not.toBeNull();
    expect(result).toMatchSnapshot();
  });

  it('renders a button with provided label', () => {
    const { getByText } = subject(defaultProps);

    const result = getByText(defaultProps.label);

    expect(result).not.toBeNull();
    expect(result).toMatchSnapshot();
  });

  it('renders a down-chevron inside dropdown', () => {
    const { container } = subject(defaultProps);
    const iconSvg = container.querySelector('svg');

    expect(iconSvg).not.toBeNull();
    expect(iconSvg).toHaveClass('tc__icon-down-chevron');
    expect(iconSvg).toMatchSnapshot();
  });

  it('opens a popup when clicked', () => {
    const { getByRole, container } = subject(defaultProps);

    const button = getByRole('button');
    fireEvent.click(button);

    const popupElements = container.getElementsByClassName('tc__dropdown__popup');

    expect(popupElements.length).toBe(1);
    expect(popupElements).toMatchSnapshot();
  });

  it('closes when clicked on while open', () => {
    const { getByRole, container } = subject(defaultProps);

    const button = getByRole('button');
    fireEvent.click(button); // open
    fireEvent.click(button); // close

    const popupElements = container.getElementsByClassName('tc__dropdown__popup');

    expect(popupElements.length).toBe(0);
    expect(popupElements).toMatchSnapshot();
  });

  it('does not close when clicked on the list', () => {
    const { getByRole, container } = subject(defaultProps);

    const button = getByRole('button');
    fireEvent.click(button);

    const popupElements = container.getElementsByClassName('tc__dropdown__popup');
    fireEvent.click(popupElements[0]);

    expect(popupElements.length).toBe(1);
    expect(popupElements).toMatchSnapshot();
  });

  it('renders up-chevron when open', () => {
    const { getByRole, container } = subject(defaultProps);

    const button = getByRole('button');
    fireEvent.click(button);

    const iconSvg = container.querySelector('svg');

    expect(iconSvg).not.toBeNull();
    expect(iconSvg).toHaveClass('tc__icon-up-chevron');
    expect(iconSvg).toMatchSnapshot();
  });

  it('renders a popup title', () => {
    const label = 'Dropdown label';
    const popupTitle = 'Dropdown popup label';

    const { getByRole, getByText } = subject({
      label,
      popupTitle,
    });

    const button = getByRole('button');
    fireEvent.click(button);

    const result = getByText(popupTitle);

    expect(result).not.toBeNull();
    expect(result).toMatchSnapshot();
  });
});
