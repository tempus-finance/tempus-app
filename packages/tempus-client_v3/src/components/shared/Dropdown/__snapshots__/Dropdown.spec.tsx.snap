// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dropdown closes when clicked on while open 1`] = `HTMLCollection []`;

exports[`Dropdown does not close when clicked on the list 1`] = `
HTMLCollection [
  <div
    class="tc__dropdown__popup"
  >
    <div
      class="tc__dropdown-popup-title"
    >
      <div
        style="font-style: normal; font-size: 10px; line-height: 16px; color: rgb(34, 34, 34); font-weight: 700; font-family: 'DM Sans', sans-serif; opacity: 1;"
      >
        Title
      </div>
    </div>
  </div>,
]
`;

exports[`Dropdown opens a popup when clicked 1`] = `
HTMLCollection [
  <div
    class="tc__dropdown__popup"
  >
    <div
      class="tc__dropdown-popup-title"
    >
      <div
        style="font-style: normal; font-size: 10px; line-height: 16px; color: rgb(34, 34, 34); font-weight: 700; font-family: 'DM Sans', sans-serif; opacity: 1;"
      >
        Title
      </div>
    </div>
  </div>,
]
`;

exports[`Dropdown renders a button element 1`] = `
<button
  class="tc__btn tc__tooltip-wrapper-anchor"
  data-selected="false"
  type="button"
>
  <div
    style="font-style: normal; font-size: 12px; line-height: 16px; color: rgb(34, 34, 34); font-weight: 400; font-family: 'DM Sans', sans-serif; opacity: 1;"
  >
    Filter
  </div>
  <svg
    class="tc__icon tc__icon-down-chevron"
    fill="none"
    height="12"
    viewBox="0 0 16 16"
    width="12"
    xmlns="http://www.w3.org/2000/svg"
  >
    <g
      clip-path="url(#clip0_909_5439)"
    >
      <path
        d="M15.7387 5.36992L8.63082 11.7636C8.54823 11.8385 8.44996 11.898 8.34169 11.9385C8.23342 11.9791 8.11729 12 8 12C7.88271 12 7.76658 11.9791 7.65831 11.9385C7.55004 11.898 7.45177 11.8385 7.36918 11.7636L0.261296 5.36992C0.0939908 5.21942 -2.49304e-09 5.01531 0 4.80248C2.49304e-09 4.58965 0.0939908 4.38554 0.261296 4.23504C0.428601 4.08455 0.655515 4 0.89212 4C1.12873 4 1.35564 4.08455 1.52294 4.23504L8 10.0693L14.4771 4.23504C14.6444 4.08455 14.8713 4 15.1079 4C15.3445 4 15.5714 4.08455 15.7387 4.23504C15.906 4.38554 16 4.58965 16 4.80248C16 5.01531 15.906 5.21942 15.7387 5.36992Z"
        fill="#222222"
      />
    </g>
    <defs>
      <clippath
        id="clip0_909_5439"
      >
        <rect
          fill="white"
          height="16"
          width="16"
        />
      </clippath>
    </defs>
  </svg>
</button>
`;

exports[`Dropdown renders a button with provided label 1`] = `
<div
  style="font-style: normal; font-size: 12px; line-height: 16px; color: rgb(34, 34, 34); font-weight: 400; font-family: 'DM Sans', sans-serif; opacity: 1;"
>
  Filter
</div>
`;

exports[`Dropdown renders a down-chevron inside dropdown 1`] = `
<svg
  class="tc__icon tc__icon-down-chevron"
  fill="none"
  height="12"
  viewBox="0 0 16 16"
  width="12"
  xmlns="http://www.w3.org/2000/svg"
>
  <g
    clip-path="url(#clip0_909_5439)"
  >
    <path
      d="M15.7387 5.36992L8.63082 11.7636C8.54823 11.8385 8.44996 11.898 8.34169 11.9385C8.23342 11.9791 8.11729 12 8 12C7.88271 12 7.76658 11.9791 7.65831 11.9385C7.55004 11.898 7.45177 11.8385 7.36918 11.7636L0.261296 5.36992C0.0939908 5.21942 -2.49304e-09 5.01531 0 4.80248C2.49304e-09 4.58965 0.0939908 4.38554 0.261296 4.23504C0.428601 4.08455 0.655515 4 0.89212 4C1.12873 4 1.35564 4.08455 1.52294 4.23504L8 10.0693L14.4771 4.23504C14.6444 4.08455 14.8713 4 15.1079 4C15.3445 4 15.5714 4.08455 15.7387 4.23504C15.906 4.38554 16 4.58965 16 4.80248C16 5.01531 15.906 5.21942 15.7387 5.36992Z"
      fill="#222222"
    />
  </g>
  <defs>
    <clippath
      id="clip0_909_5439"
    >
      <rect
        fill="white"
        height="16"
        width="16"
      />
    </clippath>
  </defs>
</svg>
`;

exports[`Dropdown renders a popup title 1`] = `
<div
  style="font-style: normal; font-size: 10px; line-height: 16px; color: rgb(34, 34, 34); font-weight: 700; font-family: 'DM Sans', sans-serif; opacity: 1;"
>
  Dropdown popup label
</div>
`;

exports[`Dropdown renders up-chevron when open 1`] = `
<svg
  class="tc__icon tc__icon-up-chevron"
  fill="none"
  height="12"
  viewBox="0 0 16 16"
  width="12"
  xmlns="http://www.w3.org/2000/svg"
>
  <g
    clip-path="url(#clip0_909_5449)"
  >
    <path
      d="M15.7372 11.7636C15.6546 11.8385 15.5564 11.898 15.4481 11.9385C15.3398 11.9791 15.2237 12 15.1064 12C14.9892 12 14.873 11.9791 14.7648 11.9385C14.6565 11.898 14.5583 11.8385 14.4757 11.7636L7.99924 5.9304L1.5228 11.7636C1.35551 11.9141 1.12862 11.9986 0.892036 11.9986C0.655453 11.9986 0.42856 11.9141 0.261271 11.7636C0.0939821 11.6132 0 11.4091 0 11.1963C0 10.9835 0.0939821 10.7794 0.261271 10.629L7.36848 4.23636C7.45107 4.16147 7.54932 4.10202 7.65758 4.06145C7.76584 4.02089 7.88196 4 7.99924 4C8.11652 4 8.23264 4.02089 8.3409 4.06145C8.44916 4.10202 8.54742 4.16147 8.63001 4.23636L15.7372 10.629C15.8205 10.7032 15.8866 10.7916 15.9317 10.889C15.9768 10.9864 16 11.0908 16 11.1963C16 11.3018 15.9768 11.4062 15.9317 11.5036C15.8866 11.601 15.8205 11.6894 15.7372 11.7636Z"
      fill="#222222"
    />
  </g>
  <defs>
    <clippath
      id="clip0_909_5449"
    >
      <rect
        fill="white"
        height="16"
        width="16"
      />
    </clippath>
  </defs>
</svg>
`;
